File: cpp/src/load_balancer.cpp
#include "moe_nexus/load_balancer.h"
#include <cmath>
#include <algorithm>
#include <numeric>

namespace moe_nexus {

LoadBalancer::LoadBalancer(int num_experts, float capacity_factor)
    : num_experts_(num_experts), capacity_factor_(capacity_factor) {
    stats_.resize(num_experts);
    for (int i = 0; i < num_experts; ++i) {
        stats_[i].expert_id = i;
    }
}

void LoadBalancer::record_routing(const RouterOutput& routing) {
    const int* indices = routing.indices.data();
    const float* scores = routing.scores.data();
    int total = static_cast<int>(routing.indices.size());
    
    for (int i = 0; i < total; ++i) {
        int expert_id = indices[i];
        if (expert_id >= 0 && expert_id < num_experts_) {
            ExpertStats& stat = stats_[expert_id];
            stat.total_tokens++;
            stat.total_weight += scores[i];
        }
    }
}

BalanceResult LoadBalancer::analyze() const {
    BalanceResult result;
    
    if (stats_.empty()) {
        return result;
    }
    
    std::vector<int> utils;
    utils.reserve(stats_.size());
    for (const auto& stat : stats_) {
        utils.push_back(stat.total_tokens);
    }
    
    float sum = 0.0f;
    for (int u : utils) sum += u;
    float mean = sum / utils.size();
    
    if (mean > 0.0f) {
        float variance = 0.0f;
        for (int u : utils) {
            float diff = u - mean;
            variance += diff * diff;
        }
        variance /= utils.size();
        float std = std::sqrt(variance);
        result.coefficient_of_variation = std / mean;
    }
    
    auto [min_it, max_it] = std::minmax_element(utils.begin(), utils.end());
    result.min_utilization = static_cast<float>(*min_it);
    result.max_utilization = static_cast<float>(*max_it);
    result.imbalance = result.max_utilization - result.min_utilization;
    
    // Generate suggestions
    if (result.coefficient_of_variation > 0.5f) {
        result.suggestions.push_back("High expert imbalance detected. Consider increasing noise during training.");
    }
    if (mean > 0.0f && result.max_utilization > mean * capacity_factor_) {
        result.suggestions.push_back("Capacity violations detected. Review expert capacity limits.");
    }
    if (mean > 0.0f && result.min_utilization < mean * 0.1f) {
        result.suggestions.push_back("Some experts are underutilized. Consider expert pruning or merging.");
    }
    
    return result;
}

void LoadBalancer::reset() {
    for (auto& stat : stats_) {
        stat.total_tokens = 0;
        stat.total_weight = 0.0f;
    }
}

} // namespace moe_nexus

File: src/moe_nexus/optimizer/load_balancer.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List

import torch


@dataclass
class ExpertStats:
    expert_id: int
    total_tokens: int = 0
    total_weight: float = 0.0
    capacity_violations: int = 0

    @property
    def avg_weight(self) -> float:
        return self.total_weight / self.total_tokens if self.total_tokens > 0 else 0.0


@dataclass
class BalanceResult:
    imbalance: float
    coefficient_of_variation: float
    max_utilization: float
    min_utilization: float
    suggestions: List[str]


class LoadBalancer:
    def __init__(self, num_experts: int, capacity_factor: float = 1.25) -> None:
        self.num_experts = num_experts
        self.capacity_factor = capacity_factor
        self._stats: Dict[int, ExpertStats] = {
            i: ExpertStats(expert_id=i) for i in range(num_experts)
        }

    def record_routing(
        self,
        topk_indices: torch.Tensor,
        topk_scores: torch.Tensor,
    ) -> None:
        flat_indices = topk_indices.reshape(-1).tolist()
        flat_scores = topk_scores.reshape(-1).tolist()

        for idx, score in zip(flat_indices, flat_scores):
            if idx >= 0 and idx < self.num_experts:
                stat = self._stats[idx]
                stat.total_tokens += 1
                stat.total_weight += score

    def analyze(self) -> BalanceResult:
        utilizations = []
        for stat in self._stats.values():
            util = stat.total_tokens
            utilizations.append(util)

        if not utilizations:
            return BalanceResult(0.0, 0.0, 0.0, 0.0, [])

        util_tensor = torch.tensor(utilizations, dtype=torch.float32)
        mean_util = util_tensor.mean().item()
        std_util = util_tensor.std().item()
        cv = std_util / mean_util if mean_util > 0 else 0.0
        max_util = util_tensor.max().item()
        min_util = util_tensor.min().item()
        imbalance = max_util - min_util

        suggestions = []
        if cv > 0.5:
            suggestions.append("High expert imbalance detected. Consider increasing noise during training.")
        if max_util > mean_util * self.capacity_factor:
            suggestions.append("Capacity violations detected. Review expert capacity limits.")
        if min_util < mean_util * 0.1:
            suggestions.append("Some experts are underutilized. Consider expert pruning or merging.")

        return BalanceResult(
            imbalance=imbalance,
            coefficient_of_variation=cv,
            max_utilization=max_util,
            min_utilization=min_util,
            suggestions=suggestions,
        )

    def reset(self) -> None:
        for stat in self._stats.values():
            stat.total_tokens = 0
            stat.total_weight = 0.0
            stat.capacity_violations = 0
